File: crates/sophia-runtime/tests/support/native_launcher_content.rs
use sophia_protocol::*;
use sophia_runtime::*;

pub const GRANT: ContentGrant = ContentGrant {
    connection_epoch: 2,
    content_grant_epoch: 2,
};
pub const OUTPUT: ContentOutputId = ContentOutputId {
    id: 2,
    generation: 3,
};
pub const ALLOCATION: ContentAllocationId = ContentAllocationId {
    id: 1,
    generation: 1,
};
pub const RESOURCE: ContentResourceId = ContentResourceId {
    id: 1,
    generation: 1,
};
pub fn tx(n: u64) -> TransactionId {
    TransactionId::from_raw(n)
}
pub fn opening() -> NativeLauncherOpening {
    NativeLauncherOpening {
        grant: GRANT,
        opening: 7,
        output: OUTPUT,
        catalog_generation: 8,
        state_revision: 1,
    }
}
pub fn catalog() -> ShellApplicationCatalog {
    ShellApplicationCatalog {
        connection_epoch: GRANT.connection_epoch,
        generation: 8,
        entries: [1, 2]
            .into_iter()
            .map(|slot| ShellApplicationDescriptor {
                slot,
                available: true,
                label: format!("app{slot}"),
                keywords: String::new(),
            })
            .collect(),
    }
}
pub fn limits() -> ContentLimits {
    let mut limits = ContentLimits::prototype(GRANT);
    limits.max_staging_bytes = 4 * 1024 * 1024;
    limits.max_resident_bytes = 12 * 1024 * 1024;
    limits.max_retiring_bytes = 8 * 1024 * 1024;
    limits
}
pub fn facts() -> ContentOutputFactsEntry {
    ContentOutputFactsEntry {
        output: OUTPUT,
        local_width: 128,
        local_height: 64,
        scale_numerator: 1,
        scale_denominator: 1,
        scale_generation: 4,
    }
}
pub fn request(id: u64) -> NativeLauncherAllocationRequest {
    NativeLauncherAllocationRequest {
        grant: GRANT,
        opening: 7,
        output: OUTPUT,
        request_id: id,
        prior: ContentAllocationId::default(),
        operation: 1,
        edge: 1,
        desired_width: 64,
        desired_height: 32,
        margins: ContentMargins::default(),
    }
}
pub fn allocation() -> ContentAllocationSnapshot {
    ContentAllocationSnapshot {
        native_opening: Some(7),
        output: OUTPUT,
        allocation: ALLOCATION,
        scale_generation: 4,
        scale_numerator: 1,
        scale_denominator: 1,
        role: 3,
        edge: 1,
        margins: ContentMargins::default(),
        logical: ContentLogicalRect {
            x: 32,
            y: 0,
            width: 64,
            height: 32,
        },
        pixel: ContentPixelRect {
            x: 32,
            y: 0,
            width: 64,
            height: 32,
        },
        parent: ContentAllocationId::default(),
        anchor_parent_rect: ContentPixelRect::default(),
        allowed_reservation_extent: 0,
    }
}
pub fn begin() -> NativeLauncherCandidateBegin {
    NativeLauncherCandidateBegin {
        content: ContentCandidateBegin {
            grant: GRANT,
            candidate_generation: 1,
            output: OUTPUT,
            facts_generation: 5,
            pacing_permit: 1,
            interaction_generation: 6,
            surface_count: 1,
            placement_count: 1,
            target_count: 2,
        },
        opening: 7,
        catalog_generation: 8,
        state_revision: 1,
        selected: 2,
        rows: vec![2, 1],
    }
}
pub fn chunk() -> ContentCandidateChunk {
    ContentCandidateChunk {
        grant: GRANT,
        candidate_generation: 1,
        chunk_ordinal: 0,
        surfaces: vec![ContentSurface {
            allocation: ALLOCATION,
            scale_generation: 4,
            role: 3,
            edge: 1,
            margins: ContentMargins::default(),
            reservation_extent: 0,
            parent_surface_index: u16::MAX,
            anchor_parent_rect: ContentPixelRect::default(),
        }],
        placements: vec![ContentPlacement {
            resource: RESOURCE,
            surface_index: 0,
            destination_x_px: 0,
            destination_y_px: 0,
        }],
        targets: [2, 1]
            .into_iter()
            .enumerate()
            .map(|(index, slot)| ContentTarget {
                surface_index: 0,
                action_kind: 2,
                target_id: index as u64 + 1,
                target_generation: 1,
                action_id: slot,
                bounds_px: ContentPixelRect {
                    x: index as i32,
                    y: 0,
                    width: 1,
                    height: 1,
                },
            })
            .collect(),
    }
}
pub fn end() -> ContentCandidateEnd {
    ContentCandidateEnd {
        grant: GRANT,
        candidate_generation: 1,
        surface_count: 1,
        placement_count: 1,
        target_count: 2,
    }
}
pub fn context(allocations: &[ContentAllocationSnapshot]) -> ContentCandidateContext<'_> {
    ContentCandidateContext {
        output: OUTPUT,
        facts_generation: 5,
        interaction_generation: 6,
        allocations,
    }
}
pub fn native(catalog: &ShellApplicationCatalog) -> NativeLauncherCandidateContext<'_> {
    NativeLauncherCandidateContext {
        opening: opening(),
        state_revision: 1,
        catalog,
    }
}
pub fn resources(registry: &mut ContentEpochRegistry) {
    upload(registry, GRANT, 255);
}
pub fn upload(registry: &mut ContentEpochRegistry, grant: ContentGrant, value: u8) {
    let store = registry.resources_mut(grant).unwrap();
    store
        .begin(
            tx(1),
            ContentResourceBegin {
                grant,
                resource: RESOURCE,
                width_px: 2,
                height_px: 1,
                rendered_scale_numerator: 1,
                rendered_scale_denominator: 1,
                pixel_format: 1,
                chunk_count: 1,
                total_bytes: 8,
            },
            0,
        )
        .unwrap();
    store
        .chunk(
            tx(2),
            &ContentResourceChunk {
                grant,
                resource: RESOURCE,
                ordinal: 0,
                offset: 0,
                bytes: vec![0, 0, value, 255, 0, 128, 0, 128],
            },
            0,
        )
        .unwrap();
    store
        .end(
            tx(3),
            &ContentResourceEnd {
                grant,
                resource: RESOURCE,
                total_bytes: 8,
                chunk_count: 1,
            },
            0,
        )
        .unwrap();
    while store.take_event().is_some() {}
}
pub fn registry() -> ContentEpochRegistry {
    registry_with_bar(false)
}
pub fn registry_with_bar(bar: bool) -> ContentEpochRegistry {
    let mut registry = ContentEpochRegistry::new(64 * 1024 * 1024).unwrap();
    if bar {
        registry
            .admit(ContentLimits::prototype(ContentGrant {
                connection_epoch: 1,
                content_grant_epoch: 1,
            }))
            .unwrap();
    }
    registry
        .admit_with_profile(limits(), ContentStoreProfile::NativeLauncher)
        .unwrap();
    let allocations = registry.allocations_mut(GRANT).unwrap();
    allocations
        .publish_outputs(tx(1), 5, vec![facts()])
        .unwrap();
    allocations.take_event().unwrap();
    resources(&mut registry);
    registry
}
pub fn grant_allocation(registry: &mut ContentEpochRegistry) -> Vec<ContentAllocationSnapshot> {
    let store = registry.allocations_mut(GRANT).unwrap();
    store
        .request_native_launcher(tx(4), request(1), opening(), 0)
        .unwrap();
    store.grant(1, allocation(), &[]).unwrap();
    store.take_event().unwrap();
    store.snapshots()
}
pub fn assemble(
    registry: &mut ContentEpochRegistry,
    allocations: &[ContentAllocationSnapshot],
    catalog: &ShellApplicationCatalog,
) {
    let (resources, store) = registry.active_parts_mut(GRANT).unwrap();
    store.grant_permit(tx(5), OUTPUT, 1, 1, 0).unwrap();
    store.take_event().unwrap();
    store
        .begin_native_launcher(tx(6), begin(), native(catalog), 0)
        .unwrap();
    store.chunk_native_launcher(tx(7), chunk(), 0).unwrap();
    store
        .end_native_launcher(
            tx(8),
            end(),
            context(allocations),
            native(catalog),
            resources,
            0,
        )
        .unwrap();
}

File: bindings/c/shell_wire/native_launcher_content.c
#include "../sophia_shell_native_launcher.h"
#include "fields.h"

static void put_identity(uint8_t *p, struct sophia_shell_native_id v)
{
    shell_put64(p, v.id);
    shell_put64(p+8, v.generation);
}
static void put_grant(uint8_t *p, struct sophia_shell_native_grant v)
{
    shell_put64(p, v.connection_epoch);
    shell_put64(p+8, v.content_grant_epoch);
}
static int encode(uint8_t *dst, size_t capacity, uint16_t kind, uint64_t tx,
                  const uint8_t *p, size_t bytes, size_t *frame_bytes)
{
    struct sophia_shell_frame f = {kind, tx, p, bytes};
    int r = sophia_shell_native_launcher_validate(&f);
    if (r != SOPHIA_SHELL_OK) return r;
    return sophia_shell_frame_encode(dst, capacity, kind, tx, p, bytes, frame_bytes);
}
int sophia_shell_native_candidate_encode(uint8_t *dst, size_t capacity,
    uint64_t tx, const struct sophia_shell_native_candidate *v, size_t *frame_bytes)
{
    if (!v) return SOPHIA_SHELL_ARGUMENT;
    if (v->row_count > SOPHIA_SHELL_NATIVE_MAX_ROWS) return SOPHIA_SHELL_INVALID;
    uint8_t p[108+2*SOPHIA_SHELL_NATIVE_MAX_ROWS] = {0};
    put_grant(p, v->grant);
    shell_put64(p+16, v->candidate_generation);
    put_identity(p+24, v->output);
    shell_put64(p+40, v->facts_generation);
    shell_put64(p+48, v->pacing_permit);
    shell_put64(p+56, v->interaction_generation);
    shell_put32(p+64, 1);
    shell_put32(p+68, v->placement_count);
    shell_put32(p+72, v->row_count);
    shell_put64(p+80, v->opening);
    shell_put64(p+88, v->catalog_generation);
    shell_put64(p+96, v->state_revision);
    shell_put16(p+104, v->selected);
    shell_put16(p+106, v->row_count);
    for (size_t i=0; i<v->row_count; ++i) shell_put16(p+108+2*i, v->rows[i]);
    return encode(dst, capacity, 189, tx, p, 108+2u*v->row_count, frame_bytes);
}
int sophia_shell_native_chunk_encode(uint8_t *dst, size_t capacity,
    uint64_t tx, const struct sophia_shell_native_chunk *v, size_t *frame_bytes)
{
    if (!v) return SOPHIA_SHELL_ARGUMENT;
    if (v->surface_count > 1 || v->placement_count > SOPHIA_SHELL_NATIVE_MAX_ROWS ||
        v->target_count > SOPHIA_SHELL_NATIVE_MAX_ROWS) return SOPHIA_SHELL_INVALID;
    uint8_t p[40+64+80*SOPHIA_SHELL_NATIVE_MAX_ROWS] = {0};
    put_grant(p, v->grant);
    shell_put64(p+16, v->candidate_generation);
    shell_put32(p+24, v->ordinal);
    shell_put32(p+28, v->surface_count);
    shell_put32(p+32, v->placement_count);
    shell_put32(p+36, v->target_count);
    size_t at = 40;
    if (v->surface_count) {
        put_identity(p+at, v->surface.allocation);
        shell_put64(p+at+16, v->surface.scale_generation);
        shell_put16(p+at+24, 3);
        shell_put16(p+at+26, v->surface.edge);
        for (size_t i=0; i<4; ++i) shell_put16(p+at+28+2*i, (uint16_t)v->surface.margins[i]);
        shell_put16(p+at+40, UINT16_MAX);
        at += 64;
    }
    for (size_t i=0; i<v->placement_count; ++i, at+=32) {
        put_identity(p+at, v->placements[i].resource);
        shell_put32(p+at+20, (uint32_t)v->placements[i].x);
        shell_put32(p+at+24, (uint32_t)v->placements[i].y);
    }
    for (size_t i=0; i<v->target_count; ++i, at+=48) {
        shell_put16(p+at+2, 2);
        shell_put64(p+at+4, v->targets[i].id);
        shell_put64(p+at+12, v->targets[i].generation);
        shell_put64(p+at+20, v->targets[i].slot);
        shell_put32(p+at+28, (uint32_t)v->targets[i].x);
        shell_put32(p+at+32, (uint32_t)v->targets[i].y);
        shell_put32(p+at+36, v->targets[i].width);
        shell_put32(p+at+40, v->targets[i].height);
    }
    return encode(dst, capacity, 190, tx, p, at, frame_bytes);
}
